File: src/main/java/com/carddemo/entity/TransactionType.java
package com.carddemo.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.util.Objects;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

/**
 * JPA entity mapping the legacy VSAM {@code TRANTYPE} key-sequenced dataset to the relational
 * table {@code transaction_type}. It is the faithful Java translation of the COBOL record
 * {@code TRAN-TYPE-RECORD} defined by copybook {@code app/cpy/CVTRA03Y.cpy} (record length 60),
 * and it preserves the two-character transaction-type code (the VSAM primary key,
 * {@code KEYLEN = 2}, {@code RKP = 0} per {@code app/catlg/LISTCAT.txt}) together with its
 * human-readable description.
 *
 * <p>This is the <strong>simplest</strong> entity in the CardDemo domain model — a single-column
 * assigned primary key and one scalar attribute — and therefore serves as the canonical example
 * of the project's entity conventions (assigned business key, {@link JdbcTypeCode} pinning for
 * {@code CHAR} columns, value-based {@link #equals(Object)}/{@link #hashCode()} on the identifier,
 * and a flat scalar-foreign-key design with no JPA associations).</p>
 *
 * <h2>Record layout (CVTRA03Y.cpy — RECLN = 60)</h2>
 * <pre>
 *   COBOL field        PIC      Java field           Column                Notes
 *   ----------------   ------   ------------------   -------------------   --------------------------
 *   TRAN-TYPE          X(02)    typeCd  (String)     type_cd  CHAR(2)      &#64;Id, assigned ("01".."07")
 *   TRAN-TYPE-DESC     X(50)    typeDesc(String)     type_desc VARCHAR(50) e.g. "Purchase"
 *   FILLER             X(08)    (not stored)         &#8212;                   COBOL padding, not persisted
 *   ----------------   ------   ------------------   -------------------   --------------------------
 *   2 + 50 + 8 = 60 bytes
 * </pre>
 *
 * <h2>Binding schema contract</h2>
 * <p>The application boots with {@code spring.jpa.hibernate.ddl-auto=validate}, so Hibernate
 * validates this mapping against the Flyway-created schema in
 * {@code src/main/resources/db/migration/V1__schema.sql}:</p>
 * <pre>
 *   CREATE TABLE transaction_type (
 *       type_cd   CHAR(2)     NOT NULL,
 *       type_desc VARCHAR(50),
 *       CONSTRAINT pk_transaction_type PRIMARY KEY (type_cd)
 *   );
 * </pre>
 * <p>Any divergence in table name, column name, or JDBC type code raises a
 * {@code SchemaManagementException} and aborts startup. In particular, {@link #typeCd} is annotated
 * with {@link JdbcTypeCode}({@link SqlTypes#CHAR}): a plain Java {@code String} otherwise maps to
 * JDBC {@code VARCHAR} (type code 12), whereas the column is {@code CHAR(2)} (type code 1), and
 * Hibernate 6.4 schema validation compares JDBC type codes — forcing {@code CHAR} keeps the entity
 * byte-compatible with the legacy fixed-width key on both H2 (PostgreSQL mode) and PostgreSQL.</p>
 *
 * <h2>Design notes</h2>
 * <ul>
 *   <li><strong>Assigned key</strong> — {@code type_cd} is a business key supplied by the seed data
 *       (V2 reference seed) and the application, never database-generated; consequently there is
 *       deliberately <em>no</em> {@code @GeneratedValue}.</li>
 *   <li><strong>Flat scalar-FK design</strong> — child tables ({@code transaction_category}, and
 *       transitively {@code transactions} / {@code transaction_category_balance}) reference this
 *       table through database-level foreign keys only; no JPA {@code @OneToMany}/{@code @ManyToOne}
 *       association is declared here, matching the project-wide repository-by-id access pattern.</li>
 *   <li><strong>Identity semantics</strong> — {@link #equals(Object)} and {@link #hashCode()} are
 *       based solely on the {@code @Id} ({@link #typeCd}), the stable identifier of the row.</li>
 * </ul>
 */
@Entity
@Table(name = "transaction_type")
public class TransactionType {

    /**
     * The two-character transaction-type code and primary key, ported from {@code TRAN-TYPE}
     * {@code PIC X(02)}. Pinned to JDBC {@code CHAR} via {@link JdbcTypeCode} so the mapping matches
     * the {@code type_cd CHAR(2)} column under {@code ddl-auto=validate}. This is an assigned
     * business key (e.g. {@code "01".."07"}); it is never generated.
     */
    @Id
    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(name = "type_cd", length = 2)
    private String typeCd;

    /**
     * The human-readable transaction-type description, ported from {@code TRAN-TYPE-DESC}
     * {@code PIC X(50)} and mapped to the {@code type_desc VARCHAR(50)} column (e.g. {@code "Purchase"}).
     */
    @Column(name = "type_desc", length = 50)
    private String typeDesc;

    /**
     * Public no-argument constructor required by the JPA specification for entity instantiation.
     */
    public TransactionType() {
        // No-arg constructor required by JPA / Hibernate.
    }

    /**
     * Convenience all-arguments constructor for tests, seeding, and mapper code.
     *
     * @param typeCd   the two-character transaction-type code (primary key)
     * @param typeDesc the transaction-type description
     */
    public TransactionType(String typeCd, String typeDesc) {
        this.typeCd = typeCd;
        this.typeDesc = typeDesc;
    }

    /**
     * Returns the two-character transaction-type code (primary key).
     *
     * @return the transaction-type code
     */
    public String getTypeCd() {
        return typeCd;
    }

    /**
     * Sets the two-character transaction-type code (primary key).
     *
     * @param typeCd the transaction-type code to set
     */
    public void setTypeCd(String typeCd) {
        this.typeCd = typeCd;
    }

    /**
     * Returns the transaction-type description.
     *
     * @return the transaction-type description
     */
    public String getTypeDesc() {
        return typeDesc;
    }

    /**
     * Sets the transaction-type description.
     *
     * @param typeDesc the transaction-type description to set
     */
    public void setTypeDesc(String typeDesc) {
        this.typeDesc = typeDesc;
    }

    /**
     * Equality is based solely on the primary key {@link #typeCd}, the stable identity of the row.
     *
     * @param o the object to compare with
     * @return {@code true} if {@code o} is a {@code TransactionType} with an equal {@code typeCd}
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof TransactionType)) {
            return false;
        }
        TransactionType that = (TransactionType) o;
        return Objects.equals(typeCd, that.typeCd);
    }

    /**
     * Hash code derived solely from the primary key {@link #typeCd}, consistent with
     * {@link #equals(Object)}.
     *
     * @return the hash code for this entity
     */
    @Override
    public int hashCode() {
        return Objects.hash(typeCd);
    }

    /**
     * Diagnostic representation including both fields; this entity carries no sensitive data.
     *
     * @return a string representation of this transaction type
     */
    @Override
    public String toString() {
        return "TransactionType{"
                + "typeCd='" + typeCd + '\''
                + ", typeDesc='" + typeDesc + '\''
                + '}';
    }
}

File: backend/app/models/transaction_type.py
# Ported from COBOL copybook CVTRA03Y (TRAN-TYPE-RECORD); VSAM TRANTYPE KSDS
# (KEYLEN=2, RKP=0). Reference-only source (never modified): app/cpy/CVTRA03Y.cpy.
# Modern SQLAlchemy 2.0 ORM port of the legacy transaction-type reference file:
# it replaces the VSAM KSDS with the PostgreSQL "transaction_type" table,
# registered on the shared app.db.base.Base metadata (AAP 0.5.1).
"""Transaction-type reference model for the CardDemo ORM layer.

Modern port of the legacy COBOL copybook ``CVTRA03Y`` (``TRAN-TYPE-RECORD``,
record length 60), which described the VSAM ``TRANTYPE`` KSDS -- a small,
standalone lookup file keyed on a two-character transaction-type code. The
copybook layout (verified against ``app/cpy/CVTRA03Y.cpy``) is::

    01  TRAN-TYPE-RECORD.
        05  TRAN-TYPE          PIC X(02).   -> primary key (RKP=0, KEYLEN=2)
        05  TRAN-TYPE-DESC     PIC X(50).
        05  FILLER             PIC X(08).   -> dropped (record-length padding)

The VSAM primary key ``TRAN-TYPE`` (relative key position 0, length 2) becomes
the PostgreSQL ``PRIMARY KEY``. The trailing ``FILLER`` is intentionally not
mapped: on the mainframe it only padded the record to its 60-byte length and
carries no business meaning. This is a reference/lookup table, so -- per the
AAP scope discipline -- it declares no foreign keys or relationships even
though other tables (for example ``transactions``) reference the type codes.

The description column is deliberately named ``tran_type_desc`` (not
``description``) so that the transaction-type Pydantic schema, which reads ORM
attributes by name through ``ConfigDict(from_attributes=True)``, maps cleanly
onto this model.
"""

from sqlalchemy import CHAR, String
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class TransactionType(Base):
    """ORM model for the ``transaction_type`` reference table.

    One row per transaction-type code, ported one-to-one from the legacy VSAM
    ``TRANTYPE`` file (copybook ``CVTRA03Y``). Column order and fixed widths are
    preserved from the COBOL record so that seed data loaded from
    ``app/data/ASCII`` and the golden-master parity checks reconcile
    field-for-field with the mainframe original.
    """

    __tablename__ = "transaction_type"

    # TRAN-TYPE PIC X(02): two-character transaction-type code and the VSAM
    # primary key (RKP=0, KEYLEN=2). Fixed-width code -> CHAR(2). The
    # primary-key constraint is auto-named ``pk_transaction_type`` by the shared
    # MetaData naming convention in app.db.base (never hand-named here).
    tran_type: Mapped[str] = mapped_column(CHAR(2), primary_key=True)

    # TRAN-TYPE-DESC PIC X(50): human-readable description of the transaction
    # type. Named ``tran_type_desc`` (never ``description``) to match the
    # transaction-type Pydantic DTO field consumed via from_attributes.
    tran_type_desc: Mapped[str] = mapped_column(String(50))

    # NOTE: the legacy copybook's trailing FILLER PIC X(08) is intentionally
    # dropped. It held no data and existed only to pad TRAN-TYPE-RECORD to its
    # 60-byte VSAM record length (2 + 50 + 8 = 60); it has no relational meaning.


__all__ = ["TransactionType"]
